File: sdk/villow-0.1.0/villow-0.1.0/src/villow/artifacts.py
from __future__ import annotations

from typing import Any

ALLOWED_ARTIFACT_TYPES = {
    "file_set",
    "structured_fields",
    "message_draft",
    "event_proposal",
    "table_data",
    "generic",
}


class Artifact:
    @staticmethod
    def file_set(
        *,
        files: list[dict[str, Any]],
        preview_urls: list[str] | None = None,
        destination_proposal: dict[str, Any] | None = None,
        title: str = "Files",
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return _artifact(
            "file_set",
            title=title,
            type_payload={"files": files},
            preview_data={"preview_urls": preview_urls or [], "file_count": len(files)},
            destination_proposal=destination_proposal,
            metadata=metadata,
        )

    @staticmethod
    def structured_fields(
        *,
        fields: dict[str, Any],
        title: str = "Structured fields",
        confidence: dict[str, float] | None = None,
    ) -> dict[str, Any]:
        return _artifact(
            "structured_fields",
            title=title,
            type_payload={"fields": fields, "confidence": confidence or {}},
            preview_data={"field_count": len(fields)},
        )

    @staticmethod
    def message_draft(
        *,
        subject: str,
        body: str,
        to: list[str] | None = None,
        cc: list[str] | None = None,
        bcc: list[str] | None = None,
        title: str = "Message draft",
    ) -> dict[str, Any]:
        return _artifact(
            "message_draft",
            title=title,
            type_payload={"to": to or [], "cc": cc or [], "bcc": bcc or [], "subject": subject, "body": body},
            preview_data={"subject": subject, "body_preview": body[:200]},
        )

    @staticmethod
    def event_proposal(
        *,
        title: str,
        start: str,
        end: str,
        attendees: list[str] | None = None,
        location: str | None = None,
    ) -> dict[str, Any]:
        return _artifact(
            "event_proposal",
            title=title,
            type_payload={"title": title, "start": start, "end": end, "attendees": attendees or [], "location": location},
            preview_data={"start": start, "end": end},
        )

    @staticmethod
    def table_data(
        *,
        columns: list[str],
        rows: list[list[Any]],
        title: str = "Table data",
    ) -> dict[str, Any]:
        return _artifact(
            "table_data",
            title=title,
            type_payload={"columns": columns, "rows": rows},
            preview_data={"column_count": len(columns), "row_count": len(rows), "sample_rows": rows[:5]},
        )

    @staticmethod
    def generic(
        *,
        payload: dict[str, Any],
        title: str = "Artifact",
        preview_data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return _artifact("generic", title=title, type_payload=payload, preview_data=preview_data or {})


def _artifact(
    artifact_type: str,
    *,
    title: str,
    type_payload: dict[str, Any],
    preview_data: dict[str, Any],
    destination_proposal: dict[str, Any] | None = None,
    metadata: dict[str, Any] | None = None,
) -> dict[str, Any]:
    if artifact_type not in ALLOWED_ARTIFACT_TYPES:
        raise ValueError(f"unsupported artifact type: {artifact_type}")
    payload: dict[str, Any] = {
        "artifact_type": artifact_type,
        "title": title,
        "type_payload": type_payload,
        "preview_data": preview_data,
    }
    if destination_proposal is not None:
        payload["destination_proposal"] = destination_proposal
    if metadata is not None:
        payload["metadata"] = metadata
    return payload

File: sdk/villow-sdk-0.1.0/package/src/artifacts.ts
export type ArtifactType = 'file_set' | 'structured_fields' | 'message_draft' | 'event_proposal' | 'table_data' | 'generic';
export type ArtifactPayload = Record<string, unknown>;

export interface VillowArtifact {
  artifact_type: ArtifactType;
  title: string;
  type_payload: ArtifactPayload;
  preview_data: ArtifactPayload;
  destination_proposal?: ArtifactPayload;
  metadata?: ArtifactPayload;
}

export class Artifact {
  static fileSet(input: {
    files: ArtifactPayload[];
    previewUrls?: string[];
    destinationProposal?: ArtifactPayload;
    title?: string;
    metadata?: ArtifactPayload;
  }): VillowArtifact {
    return artifact('file_set', {
      title: input.title ?? 'Files',
      typePayload: { files: input.files },
      previewData: { preview_urls: input.previewUrls ?? [], file_count: input.files.length },
      destinationProposal: input.destinationProposal,
      metadata: input.metadata,
    });
  }

  static structuredFields(input: { fields: ArtifactPayload; title?: string; confidence?: Record<string, number> }): VillowArtifact {
    return artifact('structured_fields', {
      title: input.title ?? 'Structured fields',
      typePayload: { fields: input.fields, confidence: input.confidence ?? {} },
      previewData: { field_count: Object.keys(input.fields).length },
    });
  }

  static messageDraft(input: {
    subject: string;
    body: string;
    to?: string[];
    cc?: string[];
    bcc?: string[];
    title?: string;
  }): VillowArtifact {
    return artifact('message_draft', {
      title: input.title ?? 'Message draft',
      typePayload: {
        to: input.to ?? [],
        cc: input.cc ?? [],
        bcc: input.bcc ?? [],
        subject: input.subject,
        body: input.body,
      },
      previewData: { subject: input.subject, body_preview: input.body.slice(0, 200) },
    });
  }

  static eventProposal(input: { title: string; start: string; end: string; attendees?: string[]; location?: string }): VillowArtifact {
    return artifact('event_proposal', {
      title: input.title,
      typePayload: {
        title: input.title,
        start: input.start,
        end: input.end,
        attendees: input.attendees ?? [],
        location: input.location,
      },
      previewData: { start: input.start, end: input.end },
    });
  }

  static tableData(input: { columns: string[]; rows: unknown[][]; title?: string }): VillowArtifact {
    return artifact('table_data', {
      title: input.title ?? 'Table data',
      typePayload: { columns: input.columns, rows: input.rows },
      previewData: { column_count: input.columns.length, row_count: input.rows.length, sample_rows: input.rows.slice(0, 5) },
    });
  }

  static generic(input: { payload: ArtifactPayload; title?: string; previewData?: ArtifactPayload }): VillowArtifact {
    return artifact('generic', {
      title: input.title ?? 'Artifact',
      typePayload: input.payload,
      previewData: input.previewData ?? {},
    });
  }
}

function artifact(
  artifactType: ArtifactType,
  input: {
    title: string;
    typePayload: ArtifactPayload;
    previewData: ArtifactPayload;
    destinationProposal?: ArtifactPayload;
    metadata?: ArtifactPayload;
  },
): VillowArtifact {
  const payload: VillowArtifact = {
    artifact_type: artifactType,
    title: input.title,
    type_payload: input.typePayload,
    preview_data: input.previewData,
  };
  if (input.destinationProposal) payload.destination_proposal = input.destinationProposal;
  if (input.metadata) payload.metadata = input.metadata;
  return payload;
}
